src/csync/csync_exclude.cpp

```diff
@@ -320,106 +322,249 @@ static CSYNC_EXCLUDE_TYPE _csync_excluded_common(c_strlist_t *excludes, const ch
+             if (match_dirs_only && filetype != CSYNC_FTW_TYPE_DIR) {
+                 match = CSYNC_NOT_EXCLUDED;
+             }
+         }
+
+         if (match == CSYNC_NOT_EXCLUDED) {
+             rc = csync_fnmatch(pattern, bname, 0);
+             if (rc == 0) {
+                 match = type;
+             }
+         }
+         if (match_dirs_only) {
+             /* restore the '/' */
+             pattern[strlen(pattern)] = '/';
+         }
+     }
 
-        type = CSYNC_FILE_EXCLUDE_LIST;
-        if (!pattern[0]) { /* empty pattern */
-            continue;
-        }
-        /* Excludes starting with ']' means it can be cleanup */
-        if (pattern[0] == ']') {
-            ++pattern;
-            if (filetype == CSYNC_FTW_TYPE_FILE) {
-                type = CSYNC_FILE_EXCLUDE_AND_REMOVE;
-            }
-        }
-        /* Check if the pattern applies to pathes only. */
-        if (pattern[strlen(pattern)-1] == '/') {
-            if (!check_leading_dirs && filetype == CSYNC_FTW_TYPE_FILE) {
-                continue;
-            }
-            match_dirs_only = true;
-            pattern[strlen(pattern)-1] = '\0'; /* Cut off the slash */
-        }
+    return match;
+}
 
-        /* check if the pattern contains a / and if, compare to the whole path */
-        if (strchr(pattern, '/')) {
-            rc = csync_fnmatch(pattern, path, FNM_PATHNAME);
-            if( rc == 0 ) {
-                match = type;
-            }
-            /* if the pattern requires a dir, but path is not, its still not excluded. */
-            if (match_dirs_only && filetype != CSYNC_FTW_TYPE_DIR) {
-                match = CSYNC_NOT_EXCLUDED;
-            }
-        }
 
-        /* if still not excluded, check each component and leading directory of the path */
-        if (match == CSYNC_NOT_EXCLUDED && check_leading_dirs) {
-            size_t j = 0;
-            if (match_dirs_only && filetype == CSYNC_FTW_TYPE_FILE) {
-                j = 1; // skip the first entry, which is bname
-            }
-            for (; j < path_components->count; ++j) {
-                rc = csync_fnmatch(pattern, path_components->vector[j], 0);
-                if (rc == 0) {
-                    match = type;
-                    break;
-                }
-            }
-        } else if (match == CSYNC_NOT_EXCLUDED && !check_leading_dirs) {
-            rc = csync_fnmatch(pattern, bname, 0);
-            if (rc == 0) {
-                match = type;
-            }
+static QString convertToFullRegexpSyntax(QString exclude)
+{
+    QString s = QRegularExpression::escape(exclude).replace("\\*", "[^/]*").replace("?", "[^/]");
+    // Anchoring to the beginning if pattern starts with a /
+    if (exclude.startsWith("/")) {
+        s.prepend("^");
+    } else {
+        // we don't want "foo/bar" to match "afoo/bar"
+        s.prepend("/");
+    }
+    if (!exclude.endsWith("/"))
+        s.append("/?");
```

Comment:
I'm pretty sure this should be `(/|$)`, otherwise the pattern "foo" will match "foobar"